### PR TITLE
Client Side Caching: Alpha support

### PR DESCRIPTION
**Pull request by dvora-h**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Do tests and lints pass with this change?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

_Please provide a description of the change here._

### CODE REVIEW

File: redis/cache.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import random
+import time
+from collections import OrderedDict, defaultdict
+
+
+class _Cache:
```
Comment (chayim) [inline redis/cache.py:6]: docstrings + documentation help

File: redis/client.py [hunk 2]
```diff
@@ -203,6 +206,13 @@ def __init__(
         redis_connect_func=None,
         credential_provider: Optional[CredentialProvider] = None,
         protocol: Optional[int] = 2,
+        client_caching: bool = False,
+        client_cache: Optional[_Cache] = None,
```
Comment (chayim) [inline redis/client.py:210]: Can we pull the default string values (lru) into the cache.py also? DEFAULT_EVICTION_POLICY

File: redis/client.py [hunk 3]
```diff
@@ -310,6 +320,12 @@ def __init__(
         else:
             self.response_callbacks.update(_RedisCallbacksRESP2)
 
+        self.client_cache = client_cache
+        self.cache_blacklist = cache_blacklist
```
Comment (chayim) [inline redis/client.py:324]: Shouldn't all these values be set only if client_cache is true? I'd assume blacklist/whitelist to be empty, when False, we can save the memory, at the very least.

File: redis/client.py [hunk 4]
```diff
@@ -525,23 +541,50 @@ def _disconnect_raise(self, conn, error):
         ):
             raise error
 
+    def get_from_local_cache(self, command):
```
Comment (chayim) [inline redis/client.py:544]: docstring, type hint
`_get_from_local_cache`

File: redis/client.py [hunk 5]
```diff
@@ -525,23 +541,50 @@ def _disconnect_raise(self, conn, error):
...
+        if (
+            self.client_cache is None
+            or command[0] in self.cache_blacklist
+            or command[0] not in self.cache_whitelist
+        ):
+            return None
+        return self.client_cache.get(command)
+
+    def add_to_local_cache(self, command, response, keys):
```
Comment (chayim) [inline redis/client.py:553]: docstring, type hint
`_add_to_local_cache`

File: redis/client.py [hunk 6]
```diff
@@ -525,23 +541,50 @@ def _disconnect_raise(self, conn, error):
...
+        if (
+            self.client_cache is not None
+            and (self.cache_blacklist == [] or command[0] not in self.cache_blacklist)
+            and (self.cache_whitelist == [] or command[0] in self.cache_whitelist)
+        ):
+            self.client_cache.set(command, response, keys)
+
+    def delete_from_local_cache(self, command):
```
Comment (chayim) [inline redis/client.py:561]: docstring, type hint
`_delete_from_local_cache_`

File: redis/client.py [hunk 7]
```diff
@@ -525,23 +541,50 @@ def _disconnect_raise(self, conn, error):
...
+        if self.client_cache is not None:
+            self.client_cache.delete(command)
```
Comment (chayim) [inline redis/client.py:563]: Is it cheaper to attempt a deletion whether or not the field exists - and just trigger the delete?

File: redis/utils.py [hunk 8]
```diff
@@ -145,3 +145,142 @@ def get_lib_version():
     except metadata.PackageNotFoundError:
         libver = "99.99.99"
     return libver
+
+
+DEFAULT_BLACKLIST = [
```
Comment (chayim) [inline redis/utils.py:150]: I think it makes more sense for this to live in cache.py

Comment (chayim) [reply redis/utils.py:150]: Also, let's sort

File: redis/cache.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    def __init__(self, max_size: int, ttl: int, eviction_policy: str, **kwargs):
+        self.max_size = max_size
+        self.ttl = ttl
+        self.eviction_policy = eviction_policy
```
Comment (chayim) [inline redis/cache.py:10]: We should have a list of policies mapped in this class. Then, map them to [enum](https://docs.python.org/3/library/enum.html) to make to make it available during configuration

File: redis/cache.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        self.cache = OrderedDict()
+        self.key_commands_map = defaultdict(set)
+        self.commands_ttl_list = []
+
+    def set(self, command, response, keys_in_command):
```
Comment (chayim) [inline redis/cache.py:15]: type hints, docstrings, etc

File: redis/cache.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        if len(self.cache) >= self.max_size:
+            self._evict()
+        self.cache[command] = {
+            "response": response,
+            "keys": keys_in_command,
+            "created_time": time.monotonic(),
```
Comment (chayim) [inline redis/cache.py:21]: `created_time` -> `ctime`? Smaller use, and ctime is accepted as created_time

File: redis/cache.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        }
+        if self.eviction_policy == "lfu":
```
Comment (chayim) [inline redis/cache.py:23]: should refer to the enum

File: redis/cache.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            self.cache[command]["access_count"] = 0
```
Comment (chayim) [inline redis/cache.py:24]: Let's declare all fields and reuse

File: redis/cache.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        self._update_key_commands_map(keys_in_command, command)
+        self.commands_ttl_list.append(command)
+
+    def get(self, command):
```
Comment (chayim) [inline redis/cache.py:28]: type hints, docstrings -> etc. let's just say for all functions

File: redis/cache.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        if command in self.cache:
+            if self._is_expired(command):
+                del self.cache[command]
+                keys_in_command = self.cache[command]["keys"]
+                self._del_key_commands_map(keys_in_command, command)
```
Comment (chayim) [inline redis/cache.py:33]: self.cache[command].get("keys") for future thread-safe use, and not worrying about `IndexError`

File: redis/cache.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+                return None
```
Comment (chayim) [inline redis/cache.py:34]: None is implied, can just use `return`

File: redis/cache.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            self._update_access(command)
+            return self.cache[command]["response"]
+        return None
```
Comment (chayim) [inline redis/cache.py:37]: None is implied, can just use `return`

File: redis/cache.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+            self.cache.move_to_end(command)
+        elif self.eviction_policy == "random":
+            pass  # Random eviction doesn't require updates
+
+    def _evict(self):
+        if self._is_expired(self.commands_ttl_list[0]):
+            self.delete(self.commands_ttl_list[0])
+        elif self.eviction_policy == "lru":
+            self.cache.popitem(last=False)
+        elif self.eviction_policy == "lfu":
```
Comment (chayim) [inline redis/cache.py:75]: enum

File: redis/cache.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+        for key in keys:
+            self.key_commands_map[key].add(command)
+
+    def _del_key_commands_map(self, keys, command):
+        for key in keys:
+            self.key_commands_map[key].remove(command)
+
+    def invalidate(self, key):
+        if key in self.key_commands_map:
+            for command in self.key_commands_map[key]:
```
Comment (chayim) [inline redis/cache.py:94]: How have if not in map return, then run the for, next?

File: redis/cache.py [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+                self.cache, key=lambda k: self.cache[k].get("access_count", 0)
+            )
+            self.cache.pop(min_access_command)
+        elif self.eviction_policy == "random":
+            random_command = random.choice(list(self.cache.keys()))
+            self.cache.pop(random_command)
+
+    def _update_key_commands_map(self, keys, command):
+        for key in keys:
+            self.key_commands_map[key].add(command)
```
Comment (chayim) [inline redis/cache.py:86]: Can we `self.key_commands_map[key] += command` or did I not understand what command is here?

File: redis/cache.py [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+
+    def _del_key_commands_map(self, keys, command):
```
Comment (chayim) [inline redis/cache.py:88]: Take a look at [itertools.filterfalse](https://docs.python.org/3.8/library/itertools.html#itertools.filterfalse), might be useful.

**Review by chayim — CHANGES_REQUESTED** [review]
@dvora-h see comments

File: redis/asyncio/client.py [hunk 22]
```diff
@@ -597,6 +597,7 @@ async def _disconnect_raise(self, conn: Connection, error: Exception):
     async def execute_command(self, *args, **options):
         """Execute a command and return a parsed response"""
         await self.initialize()
+        options.pop("keys", None)  # the keys is used only for client side caching
```
Comment (chayim) [inline redis/asyncio/client.py:600]: "the keys are used only for client side caching

File: redis/cache.py [hunk 23]
```diff
@@ -0,0 +1,326 @@
+import random
```
Comment (chayim) [inline redis/cache.py:1]: To keep in theme we discussed, maybe this should be redis.cache.LocalCache?

File: redis/cache.py [hunk 24]
```diff
@@ -0,0 +1,326 @@
...
+    "ZUNION",
+]
+
+_RESPONSE = "response"
+_KEYS = "keys"
+_CTIME = "ctime"
+_ACCESS_COUNT = "access_count"
+
+
+class EvictionPolicy(Enum):
```
Comment (chayim) [inline redis/cache.py:156]: awesome

**Comment by rueian** [comment]
Wow, this is cool. Will the client tracking be supported in this client cache?

**Comment by rueian** [comment]
> @rueian Yes, we are going to add it, this is why it's just an alpha version now I hope the next version will come soon with client tracking

Great! Can’t wait to see this powerful feature be available on redis py. This will definitely gain more acknowledgment on this feature from more developers.

**Pull request merged**